gslib/context_config.py

```diff
@@ -0,0 +1,203 @@
+  """Represents errors when provisioning a client certificate."""
+  pass
+
+
+def _IsPemSectionMarker(line):
+  """Returns (begin:bool, end:bool, name:str)."""
+  if line.startswith('-----BEGIN ') and line.endswith('-----'):
+    return True, False, line[11:-5]
+  elif line.startswith('-----END ') and line.endswith('-----'):
+    return False, True, line[9:-5]
+  else:
+    return False, False, ''
+
+
+def _SplitPemIntoSections(contents, logger):
+  """Returns dict with {name: section} by parsing contents in PEM format.
+
+  A simple parser for PEM file. Please see RFC 7468 for the format of PEM
+  file. Not using regex to improve performance catching nested matches.
+  Note: This parser requires the post-encapsulation label of a section to
+  match its pre-encapsulation label. It ignores a section without a
+  matching label.
+
+  Args:
+    contents (str): Contents of a PEM file.
+    logger (logging.logger): gsutil logger.
+
+  Returns:
+    A dict of the PEM file sections.
+  """
+  result = {}
+  pem_lines = []
+  pem_section_name = None
+
+  for line in contents.splitlines():
+    line = line.strip()
+    if not line:
+      continue
+
+    begin, end, name = _IsPemSectionMarker(line)
+    if begin:
+      if pem_section_name:
+        logger.warning('Section %s missing end line and will be ignored.' %
+                       pem_section_name)
+      if name in result.keys():
+        logger.warning('Section %s already exists, and the older section will '
+                       'be ignored.' % name)
+      pem_section_name = name
+      pem_lines = []
+    elif end:
+      if not pem_section_name:
+        logger.warning(
+            'Section %s missing a beginning line and will be ignored.' % name)
+      elif pem_section_name != name:
+        logger.warning('Section %s missing a matching end line. Found: %s' %
+                       (pem_section_name, name))
+        pem_section_name = None
+
+    if pem_section_name:
+      pem_lines.append(line)
+      if end:
+        result[name] = '\n'.join(pem_lines) + '\n'
+        pem_section_name = None
+
+  if pem_section_name:
+    logger.warning('Section %s missing an end line.' % pem_section_name)
+
+  return result
+
+
+class _ContextConfig(object):
+  """Represents the configurations associated with context aware access.
+
+  Only one instance of Config can be created for the program.
+  """
+
+  def __init__(self, logger):
+    """Initializes config.
+
+    Args:
+      logger (logging.logger): gsutil logger.
+    """
+    self.logger = logger
+
+    self.client_cert_path = None
+    self.client_cert_password = None
+    self.use_client_certificate = config.get('Credentials',
+                                             'use_client_certificate',
+                                             None) == 'True'
+
+    atexit.register(self._UnprovisionClientCert)
+
+    if self.use_client_certificate:
+      command_string = config.get('Credentials', 'cert_provider_command', None)
```

Comment:
How are we handling the case where the config gets passed down from gcloud?

Reply:
Gcloud passes them in as "-o" flags, which I think this should work for:
```
[gsutil', '-o', 'GSUtil:default_project_id=stofe-storage', '-o', 'Credentials:use_client_certificate=True', '-o', 'Credentials:cert_provider_command=abc']
```
From: http://cl/336956462